tools/garaga_rs/src/msm.rs

```diff
@@ -0,0 +1,766 @@
+use crate::{
+    ecip::{
+        core::{neg_3_base_le, run_ecip},
+        curve::CurveParamsProvider,
+        g1point::G1Point,
+        rational_function::FunctionFelt,
+    },
+    io::{
+        element_to_biguint, element_to_element, element_to_limbs, padd_function_felt,
+        parse_field_elements_from_list, parse_points_from_field_elements_list, scalar_to_limbs,
+    },
+    poseidon_transcript::CairoPoseidonTranscript,
+};
+use lambdaworks_crypto::hash::poseidon::{starknet::PoseidonCairoStark252, Poseidon};
+use lambdaworks_math::{
+    elliptic_curve::short_weierstrass::curves::{
+        bls12_381::field_extension::BLS12381PrimeField, bn_254::field_extension::BN254PrimeField,
+    },
+    field::{
+        element::FieldElement,
+        fields::fft_friendly::stark_252_prime_field::Stark252PrimeField,
+        traits::{IsPrimeField, LegendreSymbol},
+    },
+    traits::ByteConversion,
+};
+use num_bigint::{BigInt, BigUint};
+
+const CURVE_BN254: usize = 0;
+const CURVE_BLS12_381: usize = 1;
+
+pub fn msm_calldata_builder(
+    values: &[BigUint],
+    scalars: &[BigUint],
+    curve_id: usize,
+) -> Vec<BigInt> {
+    assert_eq!(values.len(), 2 * scalars.len());
+
+    if curve_id == CURVE_BN254 {
+        let elements = parse_field_elements_from_list::<BN254PrimeField>(values);
+        let points = parse_points_from_field_elements_list(&elements);
+        let n = &element_to_biguint(&BN254PrimeField::get_curve_params().n);
+        if !scalars.iter().all(|x| x < n) {
+            panic!("Scalar value must be less than the curve order");
+        }
+        return calldata_builder(&points, scalars, curve_id, true, true, false);
+    }
+
+    if curve_id == CURVE_BLS12_381 {
+        let elements = parse_field_elements_from_list::<BLS12381PrimeField>(values);
+        let points = parse_points_from_field_elements_list(&elements);
+        let n = &element_to_biguint(&BLS12381PrimeField::get_curve_params().n);
+        if !scalars.iter().all(|x| x < n) {
+            panic!("Scalar value must be less than the curve order");
+        }
+        return calldata_builder(&points, scalars, curve_id, true, true, false);
+    }
+
+    panic!("Curve ID {} not supported", curve_id);
```

Comment:
Msm should be working on all CurveID supported by garaga -- they are all Weirstrass curves and all of them should work!